pkg/config/capabilities.go

```diff
@@ -57,7 +57,153 @@ func (c *ConfigTx) GetApplicationCapabilities() (map[string]bool, error) {
+	if err != nil {
+		return err
+	}
+
+	err = addCapability(c.updated.ChannelGroup.Groups[ApplicationGroupKey], capabilities, AdminsPolicyKey, capability)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemoveChannelCapability removes capability to the provided channel config,
+// capability must be specified.
+func (c *ConfigTx) RemoveChannelCapability(capability string) error {
+	capabilities, err := c.GetChannelCapabilities()
+	if err != nil {
+		return err
+	}
+
+	err = removeCapability(c.updated.ChannelGroup, capabilities, AdminsPolicyKey, capability)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemoveOrdererCapability removes capability to the provided channel config,
+// capability must be specified.
+func (c *ConfigTx) RemoveOrdererCapability(capability string) error {
+	capabilities, err := c.GetOrdererCapabilities()
+	if err != nil {
+		return err
+	}
+
+	err = removeCapability(c.updated.ChannelGroup.Groups[OrdererGroupKey], capabilities, AdminsPolicyKey, capability)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemoveApplicationCapability removes capability to the provided channel config,
+// capability must be specified.
+func (c *ConfigTx) RemoveApplicationCapability(capability string) error {
+	capabilities, err := c.GetApplicationCapabilities()
+	if err != nil {
+		return err
+	}
+
+	err = removeCapability(c.updated.ChannelGroup.Groups[ApplicationGroupKey], capabilities, AdminsPolicyKey, capability)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// capabilitiesValue returns the config definition for a a set of capabilities.
+// It is a value for the /Channel/Orderer, Channel/Application/, and /Channel groups.
+func capabilitiesValue(capabilities []string) *standardConfigValue {
+	c := &cb.Capabilities{
+		Capabilities: make(map[string]*cb.Capability),
+	}
+
+	for _, capability := range capabilities {
+		c.Capabilities[capability] = &cb.Capability{}
+	}
+
+	return &standardConfigValue{
+		key:   CapabilitiesKey,
+		value: c,
+	}
+}
+
+func addCapability(configGroup *cb.ConfigGroup, capabilities []string, modPolicy string, capability string) error {
+	for _, c := range capabilities {
+		if c == capability {
+			return errors.New("capability already exists")
+		}
+	}
+
+	err := addValue(configGroup, capabilitiesValue([]string{capability}), modPolicy)
+	if err != nil {
+		return fmt.Errorf("adding capability: %v", err)
+	}
+
+	return nil
+}
+
+func removeCapability(configGroup *cb.ConfigGroup, capabilities []string, modPolicy string, capability string) error {
+	for i, c := range capabilities {
+		if c == capability {
+			// remove the capability from slice
+			capabilities = append(capabilities[:i], capabilities[i:]...)
+
+			err := addValue(configGroup, capabilitiesValue(capabilities), modPolicy)
+			if err != nil {
+				return fmt.Errorf("removing capability: %v", err)
+			}
+
+			return nil
+		}
+	}
+
+	return errors.New("capability not set")
+}
```

Comment:
Since duplicates are not expected, this works but it makes me uneasy to see a removal during iteration. I'd much rather see an approach that does not rely on unique values.

Also: https://github.com/golang/go/wiki/SliceTricks

Reply:
🛠